libass/ass_render.c

```diff
@@ -2252,91 +2263,79 @@ size_t ass_composite_construct(void *key, void *value, void *priv)
     BitmapRef *last = NULL, *last_o = NULL;
     for (int i = 0; i < k->bitmap_count; i++) {
         BitmapRef *ref = &k->bitmaps[i];
-        if (ref->image && ref->image->bm) {
-            rectangle_combine(&rect, ref->image->bm, ref->pos.x, ref->pos.y);
+        if (ref->bm) {
+            rectangle_combine(&rect, ref->bm, ref->pos);
             last = ref;
             n_bm++;
         }
-        if (ref->image_o && ref->image_o->bm) {
-            rectangle_combine(&rect_o, ref->image_o->bm, ref->pos_o.x, ref->pos_o.y);
+        if (ref->bm_o) {
+            rectangle_combine(&rect_o, ref->bm_o, ref->pos_o);
             last_o = ref;
             n_bm_o++;
         }
     }
 
     int bord = be_padding(k->filter.be);
     if (!bord && n_bm == 1) {
-        v->bm = copy_bitmap(render_priv->engine, last->image->bm);
-        if (v->bm) {
-            v->bm->left += last->pos.x;
-            v->bm->top  += last->pos.y;
-        }
-    } else if (n_bm) {
-        v->bm = alloc_bitmap(render_priv->engine,
-                             rect.x_max - rect.x_min + 2 * bord,
-                             rect.y_max - rect.y_min + 2 * bord, true);
-        Bitmap *dst = v->bm;
-        if (dst) {
-            dst->left = rect.x_min - bord;
-            dst->top  = rect.y_min - bord;
-            for (int i = 0; i < k->bitmap_count; i++) {
-                if (!k->bitmaps[i].image)
-                    continue;
-                Bitmap *src = k->bitmaps[i].image->bm;
-                if (!src)
-                    continue;
-                int x = k->bitmaps[i].pos.x + src->left - dst->left;
-                int y = k->bitmaps[i].pos.y + src->top  - dst->top;
-                assert(x >= 0 && x + src->w <= dst->w);
-                assert(y >= 0 && y + src->h <= dst->h);
-                unsigned char *buf = dst->buffer + y * dst->stride + x;
-                render_priv->engine->add_bitmaps(buf, dst->stride,
-                                                 src->buffer, src->stride,
-                                                 src->h, src->w);
-            }
+        copy_bitmap(render_priv->engine, &v->bm, last->bm);
+        v->bm.left += last->pos.x;
+        v->bm.top  += last->pos.y;
```

Comment:
I don't want to add `buffer` check around coordinate corrections like that. So I explicitly set coordinates to zero on allocation failure to protect from UB on integer overflow due to some garbage values.

Reply:
That’s precisely why I’m telling you that _here,_ you actually set the coordinates to nonzero values even when the bitmap fails to allocate. Want to make these two lines conditional on the success of the `copy_bitmap`?

Reply:
It's because I don't want to make such lines conditional that I `memset` the whole Bitmap to zero. Coordinates for NULL bitmaps are not used anyway so I don't care about exact value and my only concern is UB due to large previous value before `copy_bitmap`.